src/app/pages/private/profile/profile.component.ts

```diff
@@ -1,29 +1,119 @@
+      role: new ExtendedFormControl('', [Validators.required], 'role'),
+      className: 'profile'
+    })
+
+    this.resetPwdForm = this.fb.group({
+      current_password: new ExtendedFormControl('', [Validators.required, Validators.minLength(PASSWORD_LENGTH)], 'current_password'),
+      new_password: new ExtendedFormControl('', [Validators.required, Validators.minLength(PASSWORD_LENGTH)], 'new_password'),
+      confirm_password: new ExtendedFormControl('', [Validators.required, Validators.minLength(PASSWORD_LENGTH), this.validatePassword], 'confirm_password'),
+      className: 'reset-password'
+    })
+  }
+
+  validatePassword(control: AbstractControl): { [key: string]: any } | null {
+    if (control.value && control.root.get('new_password') && control.value != control.root.get('new_password').value) {
+      return {'passwordNotEqual': 'Confirm Password doesn\'t match'};
+    }
   }
 
   ngOnInit(): void {
-    this.name = 'Logesh';
-    this.mobile = '9176786586';
-    this.email = 'ddlogesh@gmail.com';
-    this.enable = true;
-    this.selectedRole = 'Admin';
+    this.profileService.getProfile()
+      .then(response => {
+        this.profileForm.get('name').setValue(response['data']['name']);
+        this.profileForm.get('mobile').setValue("+91 " + response['data']['mobile']);
+        this.profileForm.get('email').setValue(response['data']['email']);
+        this.profileForm.get('two_fa_enabled').setValue(response['data']['two_fa_enabled']);
+        this.profileForm.get('role').setValue('Employee');
+      })
+      .catch(error => {
+        handleError(error, this.profileForm)
+      })
+      .finally(() => {
+        $('form.profile div.form-group-mobile input').attr('readonly', true);
+        $('form.profile div.form-group-email input').attr('readonly', true);
+        $('form.profile div.form-group-role input').attr('readonly', true);
+        this.profileApiResponse = this.profileForm.value;
+      })
+  }
+
+  updatePassword() {
+    this.profileService.updatePassword(this.resetPwdForm.get('current_password').value, this.resetPwdForm.get('new_password').value, this.resetPwdForm.get('confirm_password').value)
+      .then(response => {
+        this.profileService.logout();
+      })
+      .catch(error => {
+        handleError(error, this.resetPwdForm)
+      });
+  }
+
+  updateTwoFa() {
+    this.profileService.updateProfile({two_fa_enabled: this.profileForm.get('two_fa_enabled').value})
+      .then(response => {
+        this.profileForm.get('two_fa_enabled').setValue(response['data']['two_fa_enabled']);
+        this.profileApiResponse = this.profileForm.value;
+        if (this.profileForm.get('two_fa_enabled').value == true) {
+          this.profileService.logout();
+        }
+      })
+      .catch(error => {
+        handleError(error, this.profileForm);
+    })
+  }
+
+  updateProfile() {
+    let requestObj = {};
+    Object.keys(this.profileForm.value).map(k => {
```

Comment:
Can you try hash1.to_a - hash2.to_a & get difference?
https://stackoverflow.com/a/4928824/10489902